crates/pallet-domains/src/lib.rs

```diff
@@ -364,27 +423,72 @@ mod pallet {
 
             Ok(())
         }
+
+        #[pallet::call_index(4)]
+        #[pallet::weight((Weight::from_all(10_000), Pays::Yes))]
+        // TODO: proper benchmark
+        pub fn instantiate_domain(
+            origin: OriginFor<T>,
+            domain_config: DomainConfig,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            can_instantiate_domain::<T>(&who, &domain_config).map_err(Error::<T>::from)?;
+
+            let runtime_id = domain_config.runtime_id;
+            let created_at = frame_system::Pallet::<T>::current_block_number();
+
+            let domain_id = do_instantiate_domain::<T>(domain_config, who, created_at)
```

Comment:
IMO, do_* should internally call can_* as do_* seems to return an error irrespective